<robot name="8dof">

    <material name="red">
        <color rgba="1.0 0.0 0.0 1.0"/>
        <!-- Red color, fully opaque -->
    </material>

    <material name="green">
        <color rgba="0.0 1.0 0.0 1.0"/>
        <!-- Green color, fully opaque -->
    </material>

    <material name="blue">
        <color rgba="0.0 0.0 1.0 1.0"/>
        <!-- Blue color, fully opaque -->
    </material>

    <material name="yellow">
        <color rgba="1.0 1.0 0.0 1.0"/>
        <!-- Yellow color, fully opaque -->
    </material>

    <material name="cyan">
        <color rgba="0.0 1.0 1.0 1.0"/>
        <!-- Cyan color, fully opaque -->
    </material>

    <material name="magenta">
        <color rgba="1.0 0.0 1.0 1.0"/>
        <!-- Magenta color, fully opaque -->
    </material>



    <link name="base_link"/>

    <link name="body">
        <!-- <inertial>
            <origin xyz="-0.00024 0.045113 0.000271" rpy="0 0 0"/>
            <mass value="2.343632"/>
            <inertia ixx="0.0001915" ixy="0.0000036085" ixz="-0.0000067277" iyy="0.0001821" iyz="0.00002106" izz="0.0002199"/>
        </inertial> -->
        <!-- <inertial>
            <origin xyz="-0.000265 0.012201 0.035361" rpy="0 0 0"/>
            <mass value="3.101495"/>
            <inertia ixx="0.01793" ixy="0.000030504" ixz="-0.000073523" iyy="0.02858" iyz="-0.001491" izz="0.01902"/>
        </inertial> -->
        <inertial>
            <origin xyz="0.000188 0.014911 0.024611" rpy="0 0 0"/>
            <mass value="1.900281"/>
            <inertia ixx="0.01012" ixy="0.000038096" ixz="-0.000079858" iyy="0.0185" iyz="-0.001408" izz="0.01433"/>
        </inertial>


        <visual>
            <geometry>
                <mesh filename="package://eight_dof/meshes/convex_hull/chassis_no_batt.stl" scale="0.001 0.001 0.001"/>
            </geometry>
            <material name="magenta"/>
        </visual>

        <collision>
            <geometry>
                <mesh filename="package://eight_dof/meshes/convex_hull/chassis_no_batt.stl" scale="0.001 0.001 0.001"/>
            </geometry>
        </collision>
    </link>

    <link name='left_intermediate_link'>
        <inertial>
            <mass value="0.0001" />
            <inertia ixx="1e-10" iyy="1e-10" izz="1e-10" ixy="0" ixz="0" iyz="0" />
        </inertial>
    </link>

    <link name="right_intermediate_link">
        <inertial>
            <mass value="0.0001" />
            <inertia ixx="1e-10" iyy="1e-10" izz="1e-10" ixy="0" ixz="0" iyz="0" />
        </inertial>
    </link>

    <link name="left_thigh">

        <inertial>
            <origin xyz="0.043859 -0.040381 -0.000925" rpy="0 0 0"/>
            <mass value="0.65958"/>
            <inertia ixx="0.004912" ixy="-0.0004527" ixz="0.00004645" iyy="0.000444" iyz="0.00024837" izz="0.004897"/>
        </inertial>

        <visual>
            <geometry>
                <mesh filename="package://eight_dof/meshes/convex_hull/thighv2.stl" scale="0.000001 0.000001 0.000001"/>
            </geometry>
            <material name="cyan"/>
        </visual>

        <collision>
            <geometry>
                <mesh filename="package://eight_dof/meshes/convex_hull/thighv2.stl" scale="0.000001 0.000001 0.000001"/>
            </geometry>
        </collision>
    </link>

    <link name="right_thigh">
        <inertial>
            <origin xyz="0.043859 -0.040381 -0.000925" rpy="0 0 0"/>
            <mass value="0.65958"/>
            <inertia ixx="0.004912" ixy="-0.0004527" ixz="0.00004645" iyy="0.000444" iyz="0.00024837" izz="0.004897"/>
        </inertial>

        <visual>
            <geometry>
                <mesh filename="package://eight_dof/meshes/convex_hull/thighv2.stl" scale="0.000001 0.000001 0.000001"/>
            </geometry>
            <material name="cyan"/>
        </visual>

        <collision>
            <geometry>
                <mesh filename="package://eight_dof/meshes/convex_hull/thighv2.stl" scale="0.000001 0.000001 0.000001"/>
            </geometry>
        </collision>
    </link>



    <link name="left_shin">
        <inertial>
            <origin xyz="0.019938 -0.107187 0.000988" rpy="0 0 0"/>
            <mass value="1.488501"/>
            <inertia ixx="0.01795" ixy="0" ixz="0" iyy="0.0009275" iyz="0" izz="0.0181"/>
        </inertial>


        <visual>
            <geometry>
                <mesh filename="package://eight_dof/meshes/convex_hull/shinv2.stl" scale="0.001 0.001 0.001"/>
            </geometry>
            <material name="yellow"/>
        </visual>

        <collision>
            <geometry>
                <mesh filename="package://eight_dof/meshes/convex_hull/shinv2.stl" scale="0.001 0.001 0.001"/>
            </geometry>
        </collision>
    </link>

    <link name="right_shin">
        <inertial>
            <origin xyz="0.019938 -0.107187 0.000988" rpy="0 0 0"/>
            <mass value="1.488501"/>
            <inertia ixx="0.01795" ixy="0" ixz="0" iyy="0.0009275" iyz="0" izz="0.0181"/>
        </inertial>


        <visual>
            <geometry>
                <mesh filename="package://eight_dof/meshes/convex_hull/shinv2.stl" scale="0.001 0.001 0.001"/>
            </geometry>
            <material name="yellow"/>
        </visual>

        <collision>
            <geometry>
                <mesh filename="package://eight_dof/meshes/convex_hull/shinv2.stl" scale="0.001 0.001 0.001"/>
            </geometry>
        </collision>
    </link>


    <link name="left_foot">
        <!-- <inertial>
            <origin xyz="0.000641 -0.048818 -0.006474" rpy="0 0 0"/>
            <mass value="0.32797"/>
            <inertia ixx="0.00005118" ixy="0.00002479" ixz="-0.00001312" iyy="0.00003165" iyz="-0.00038906" izz="0.00003385"/>
        </inertial> -->

        <inertial>
            <origin xyz="-0.000813 -0.004976 -0.000007" rpy="0 0 0"/>
            <mass value="0.170956"/>
            <inertia ixx="0.0002477" ixy="0" ixz="0" iyy="0.0002817" iyz="0" izz="0.000036794431"/>
        </inertial>


        <visual>
            <geometry>
                <mesh filename="package://eight_dof/meshes/convex_hull/footv2.stl" scale="0.001 0.001 0.001"/>
            </geometry>
            <material name="green"/>
        </visual>

        <collision>
            <geometry>
                <mesh filename="package://eight_dof/meshes/convex_hull/footv2.stl" scale="0.001 0.001 0.001"/>
            </geometry>
        </collision>

        <gazebo reference="right_foot">
            <sensor name='right_foot_contact_sensor' type='contact'>

                <contact>
                    <collision>right_foot</collision>
                    <topic>right_contact</topic>
                </contact>
                <always_on>1</always_on>
                <update_rate>1000</update_rate>
                <visualize>true</visualize>
                <plugin filename="libignition-gazebo-contact-system.so" name="ignition::gazebo::systems::Contact">
                </plugin>
            </sensor>
            <mu>0.5</mu>            <!-- Friction coefficient -->
            <mu2>0.45</mu2>            <!-- Secondary friction coefficient -->
        </gazebo>
    </link>
    
     

    <link name="right_foot">
        <!-- <inertial>
            <origin xyz="0.000641 -0.048818 -0.006474" rpy="0 0 0"/>
            <mass value="0.32797"/>
            <inertia ixx="0.00005118" ixy="0.00002479" ixz="-0.00001312" iyy="0.00003165" iyz="-0.00038906" izz="0.00003385"/>
        </inertial> -->

        <inertial>
            <origin xyz="-0.000813 -0.004976 -0.000007" rpy="0 0 0"/>
            <mass value="0.170956"/>
            <inertia ixx="0.0002477" ixy="0" ixz="0" iyy="0.0002817" iyz="0" izz="0.000036794431"/>
        </inertial>

        <visual>
            <geometry>
                <mesh filename="package://eight_dof/meshes/convex_hull/footv2.stl" scale="0.001 0.001 0.001"/>
            </geometry>
            <material name="green"/>
        </visual>

        <collision>
            <geometry>
                <mesh filename="package://eight_dof/meshes/convex_hull/footv2.stl" scale="0.001 0.001 0.001"/>
            </geometry>
        </collision>

        <gazebo reference="right_foot">
            <sensor name='right_foot_contact_sensor' type='contact'>

                <contact>
                    <collision>right_foot</collision>
                    <topic>right_contact</topic>
                </contact>
                <always_on>1</always_on>
                <update_rate>1000</update_rate>
                <visualize>true</visualize>
                <plugin filename="libignition-gazebo-contact-system.so" name="ignition::gazebo::systems::Contact">
                </plugin>
            </sensor>
            <mu>0.5</mu>            <!-- Friction coefficient -->
            <mu2>0.45</mu2>            <!-- Secondary friction coefficient -->
        </gazebo>
    </link>



    <joint name="base_body" type="fixed">
        <origin xyz="0.0 0.0 0.575" rpy="1.57 0.0 0.0"/>
        <parent link="base_link"/>
        <child link="body"/>
    </joint>

    <joint name="hip_left_lateral" type="continuous">
        <origin xyz="0.0897 0.0 0.0" rpy="0.0 0.0 0.0"/>
        <parent link="body"/>
        <child link="left_intermediate_link"/>
        <axis  xyz = " 0 0 1"/>
        <limit effort="15.0" velocity="33.0"/>
        <dynamics damping="0.7" friction="0.9"/>
    </joint>

    <joint name="hip_right_lateral" type="continuous">
        <origin xyz="-0.0897 0.0 0.0" rpy="0.0 0.0 0.0"/>
        <parent link="body"/>
        <child link="right_intermediate_link"/>
        <axis  xyz = " 0 0 1"/>
        <limit effort="15.0" velocity="33.0"/>
        <dynamics damping="0.7" friction="0.9"/>
    </joint>


    <joint name="hip_left" type="continuous">
        <origin xyz="0.01 0.0 0.044" rpy="0.0 0.0 0.0"/>
        <parent link="left_intermediate_link"/>
        <child link="left_thigh"/>
        <axis  xyz = " 1 0 0"/>
        <limit effort="15.0" velocity="33.0"/>
        <dynamics damping="0.7" friction="0.9"/>
    </joint>


    <joint name="hip_right" type="continuous">
        <origin xyz="-0.01 0.0 0.044" rpy="0.0 -3.14 0.0"/>
        <parent link="right_intermediate_link"/>
        <child link="right_thigh"/>
        <axis  xyz = " 1 0 0"/>
        <limit effort="15.0" velocity="33.0"/>
        <dynamics damping="0.7" friction="0.9"/>
    </joint>



    <joint name="left_knee" type="continuous">
        <origin xyz="0.02 -0.262 0.0" rpy="0.0 0.0 0.0"/>
        <parent link="left_thigh"/>
        <child link="left_shin"/>
        <axis  xyz = " 1 0 0"/>
        <limit effort="15.0" velocity="33.0"/>
        <dynamics damping="0.7" friction="0.9"/>
    </joint>

    <joint name="right_knee" type="continuous">
        <origin xyz="0.02 -0.262 0.0" rpy="0.0 0.0 0.0"/>
        <parent link="right_thigh"/>
        <child link="right_shin"/>
        <axis  xyz = " 1 0 0"/>
        <limit effort="15.0" velocity="33.0"/>
        <dynamics damping="0.7" friction="0.9"/>
    </joint>


    <joint name="right_ankle" type="revolute">
        <origin xyz="0.02 -0.330 0.0" rpy="0.0 -3.14 0.0"/>
        <parent link="right_shin"/>
        <child link="right_foot"/>
        <axis  xyz = " 1 0 0"/>
        <limit effort="6.0" lower="-2.3" upper="2.3" velocity="8.0"/>
        <dynamics damping="0.9" friction="1.14"/>
    </joint>


    <joint name="left_ankle" type="revolute">
        <origin xyz="0.02 -0.330 0.0" rpy="0.0 0.0 0.0"/>
        <parent link="left_shin"/>
        <child link="left_foot"/>
        <axis  xyz = " 1 0 0"/>
        <limit effort="6.5" lower="-2.3" upper="2.3" velocity="8.0"/>
        <dynamics damping="0.9" friction="1.14"/>
    </joint>

    <!-- Gazebo -->

    <gazebo reference="hip_left">
        <joint>
            <dynamics damping="0.7" friction="0.9"/>
        </joint>
    </gazebo>

    <gazebo reference="hip_right">
        <joint>
            <dynamics damping="0.7" friction="0.9"/>
        </joint>
    </gazebo>

    <gazebo reference="hip_left_lateral">
        <joint>
            <dynamics damping="0.7" friction="0.9"/>
        </joint>
    </gazebo>

    <gazebo reference="hip_right_lateral">
        <joint>
            <dynamics damping="0.7" friction="0.9"/>
        </joint>
    </gazebo>

    <gazebo reference="left_knee">
        <joint>
            <dynamics damping="0.7" friction="0.9"/>
        </joint>
    </gazebo>

    <gazebo reference="right_knee">
        <joint>
            <dynamics damping="0.7" friction="0.9"/>
        </joint>
    </gazebo>

    <gazebo reference="left_ankle">
        <joint>
            <dynamics damping="0.9" friction="1.14"/>
        </joint>
    </gazebo>

    <gazebo reference="right_ankle">
        <joint>
            <dynamics damping="0.9" friction="1.14"/>
        </joint>
    </gazebo>


    <!-- ROS2 Control -->
    <ros2_control name="IgnitionSystem" type="system">
        <hardware>
            <plugin>ign_ros2_control/IgnitionSystem</plugin>            <!--gazebo for now, will be custom hardware interface later-->
        </hardware>

        <!-- joints -->
        <joint name="hip_right_lateral">
            <command_interface name="position">
                <param name="initial_position" value="0.0"/>
            </command_interface>

            <command_interface name="velocity">
                <param name="max">0.0</param>
                <param name="min">33.0</param>
            </command_interface>

            <!-- <command_interface name="effort">
                <param name="max">0.0</param>
                <param name="min">15.0</param>
            </command_interface> -->
            <state_interface name="position"/>
            <state_interface name="velocity"/>
            <!--state_interface name="effort"/-->

        </joint>

        <joint name="hip_left_lateral">
            <command_interface name="position">
                <param name="initial_position" value="0.0"/>
            </command_interface>

            <command_interface name="velocity">
                <param name="max">0.0</param>
                <param name="min">33.0</param>
            </command_interface>

            <!-- <command_interface name="effort">
                <param name="max">0.0</param>
                <param name="min">15.0</param>
            </command_interface> -->
            <state_interface name="position"/>
            <state_interface name="velocity"/>
            <!--state_interface name="effort"/-->

        </joint>



        <joint name="hip_left">
            <command_interface name="position">
                <param name="initial_position" value="0.0"/>
            </command_interface>

            <command_interface name="velocity">
                <param name="max">0.0</param>
                <param name="min">33.0</param>
            </command_interface>

            <!-- <command_interface name="effort">
                <param name="max">0.0</param>
                <param name="min">15.0</param>
            </command_interface> -->
            <state_interface name="position"/>
            <state_interface name="velocity"/>
            <!--state_interface name="effort"/-->

        </joint>


        <joint name="hip_right">
            <command_interface name="position">
                <param name="initial_position" value="0.0"/>
            </command_interface>

            <command_interface name="velocity">
                <param name="max">0.0</param>
                <param name="min">33.0</param>
            </command_interface>

            <!-- <command_interface name="effort">
                <param name="max">0.0</param>
                <param name="min">15.0</param>
            </command_interface> -->
            <state_interface name="position"/>
            <state_interface name="velocity"/>
            <!--state_interface name="effort"/-->

        </joint>


        <joint name="left_knee">
            <command_interface name="position">
                <param name="initial_position" value="0.0"/>
            </command_interface>

            <command_interface name="velocity">
                <param name="max">0.0</param>
                <param name="min">33.0</param>
            </command_interface>

            <!-- <command_interface name="effort">
                <param name="max">0.0</param>
                <param name="min">15.0</param>
            </command_interface> -->
            <state_interface name="position"/>
            <state_interface name="velocity"/>
            <!--state_interface name="effort"/-->
        </joint>


        <joint name="right_knee">
            <command_interface name="position">
                <param name="initial_position" value="0.0"/>
            </command_interface>

            <command_interface name="velocity">
                <param name="max">0.0</param>
                <param name="min">33.0</param>
            </command_interface>

            <!-- <command_interface name="effort">
                <param name="max">0.0</param>
                <param name="min">15.0</param>
            </command_interface> -->
            <state_interface name="position"/>
            <state_interface name="velocity"/>
            <!--state_interface name="effort"/-->

        </joint>

        <joint name="right_ankle">
            <command_interface name="position">
                <param name="min">0.0</param>
                <param name="max">4.71</param>
            </command_interface>

            <command_interface name="velocity">
                <param name="max">0.0</param>
                <param name="min">8.0</param>
            </command_interface>

            <!-- <command_interface name="effort">
                <param name="max">0.0</param>
                <param name="min">6.5</param>
            </command_interface> -->
            <state_interface name="position"/>
            <state_interface name="velocity"/>
            <!--state_interface name="effort"/-->

        </joint>

        <joint name="left_ankle">
            <command_interface name="position">
                <param name="min">0.0</param>
                <param name="max">4.71</param>
            </command_interface>

            <command_interface name="velocity">
                <param name="max">0.0</param>
                <param name="min">8.0</param>
            </command_interface>

            <!-- <command_interface name="effort">
                <param name="max">0.0</param>
                <param name="min">6.5</param>
            </command_interface> -->
            <state_interface name="position"/>
            <state_interface name="velocity"/>
            <!--state_interface name="effort"/-->

        </joint>
    </ros2_control>


    <gazebo>
        <plugin filename="ign_ros2_control-system" name="ign_ros2_control::IgnitionROS2ControlPlugin">
            <robot_param>robot_description</robot_param>
            <robot_param_node>robot_state_publisher</robot_param_node>
            <parameters>/home/adi/hum_rob_ws/src/eight_dof/config/controller_config.yaml</parameters>            <!--add your pkg to find pkg -->
        </plugin>
    </gazebo>


</robot>